File: competition-serv-service/src/main/java/compman/compsrv/service/schedule/ScheduleProducer.kt
package compman.compsrv.service.schedule

import arrow.core.Tuple3
import compman.compsrv.model.dto.dashboard.MatDescriptionDTO
import compman.compsrv.model.dto.schedule.*
import compman.compsrv.service.schedule.internal.InternalFightStartTime
import compman.compsrv.service.schedule.internal.InternalMatScheduleContainer
import compman.compsrv.service.schedule.internal.ScheduleAccumulator
import compman.compsrv.util.IDGenerator
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.math.BigDecimal
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.util.*
import java.util.function.Supplier

class ScheduleProducer(val competitionId: String,
                       val startTime: Map<String, Instant>,
                       val mats: List<MatDescriptionDTO>,
                       private val req: List<ScheduleRequirementDTO>,
                       private val stages: StageGraph,
                       private val periods: List<PeriodDTO>,
                       val timeBetweenFights: Map<String, BigDecimal>,
                       riskFactor: Map<String, BigDecimal>,
                       val timeZone: String) {

    companion object {
        private val log: Logger = LoggerFactory.getLogger(ScheduleProducer::class.java)
    }

    private val pauses = req.filter { it.entryType == ScheduleRequirementType.FIXED_PAUSE }.groupBy { it.matId }.mapValues { e -> e.value.sortedBy { it.startTime.toEpochMilli() }.toMutableList() }

    private val riskCoeff = riskFactor.mapValues { BigDecimal.ONE.plus(it.value) }

    private fun eightyPercentOfDurationInMillis(duration: BigDecimal): Long {
        return duration.multiply(BigDecimal.valueOf(0.8 * 60000)).toLong()
    }


    private fun createPauseEntry(pauseReq: ScheduleRequirementDTO, startTime: Instant, endTime: Instant, pauseType: ScheduleEntryType): ScheduleEntryDTO {
        return ScheduleEntryDTO().apply {
            id = pauseReq.id
            categoryIds = emptyArray()
            fightIds = arrayOf(MatIdAndSomeId().setSomeId(pauseReq.id).setMatId(pauseReq.matId!!).setStartTime(startTime))
            periodId = pauseReq.periodId
            this.startTime = startTime
            numberOfFights = 0
            entryType = pauseType
            this.endTime = endTime
            duration = pauseReq.durationMinutes
            requirementIds = arrayOf(pauseReq.id)
        }
    }

    private fun createFixedPauseEntry(fixedPause: ScheduleRequirementDTO, endTime: Instant) = createPauseEntry(fixedPause, fixedPause.startTime, endTime, ScheduleEntryType.FIXED_PAUSE)

    private fun createRelativePauseEntry(requirement: ScheduleRequirementDTO, startTime: Instant, endTime: Instant) = createPauseEntry(requirement, startTime, endTime, ScheduleEntryType.RELATIVE_PAUSE)



    private fun getFightDuration(duration: BigDecimal, periodId: String) = duration.multiply(riskCoeff[periodId]
            ?: error("No risk coeff for $periodId")).plus(timeBetweenFights[periodId] ?: BigDecimal.ZERO)

    fun simulate(): Tuple3<List<ScheduleEntryDTO>, List<InternalMatScheduleContainer>, Set<String>> {
        return this.stages
                .let { st ->
                    var fightsDispatched = 0
                    val scheduleRequirements = req
                    val unfinishedRequirements = LinkedList<ScheduleRequirementDTO>()
                    val accumulator = supplier().get()
                    val categoryIdsToFightIds = st.getCategoryIdsToFightIds()
                    val matsToIds = accumulator.matSchedules.map { it.id to it }.toMap()
                    val sortedPeriods = periods.sortedBy { it.startTime }
                    val requiremetsGraph = RequirementsGraph(scheduleRequirements.map { it.id to it }.toMap(), categoryIdsToFightIds, sortedPeriods.map { it.id }.toTypedArray())
                    val requirementsCapacity = requiremetsGraph.getRequirementsFightsSize()
                    sortedPeriods.forEach { period ->
                        val periodMats = accumulator.matSchedules.filter { it.periodId == period.id }
                        unfinishedRequirements.forEach { r ->
                            log.error("Requirement ${r.id} is from period ${r.periodId} but it is processed in ${period.id} because it could not have been processed in it's own period (wrong order)")
                            st.flushNonCompletedFights(requiremetsGraph.getFightIdsForRequirement(r.id)).forEach {
                                log.warn("Marking fight $it as invalid")
                                accumulator.invalidFights.add(it)
                            }
                        }
                        val q: Queue<Pair<ScheduleRequirementDTO, List<String>>> = LinkedList()
                        val rq: Queue<ScheduleRequirementDTO> = LinkedList(requiremetsGraph.orderedRequirements.filter { it.periodId == period.id })

                        while (!rq.isEmpty() || !unfinishedRequirements.isEmpty()) {
                            val n = st.getNonCompleteCount()
                            val onlyUnfinished = rq.isEmpty()
                            var i = 0
                            while((!rq.isEmpty() && i < initialFightsByMats.size) || !unfinishedRequirements.isEmpty()) {
                                val sr= unfinishedRequirements.poll() ?: run {
                                    i++
                                    rq.poll()
                                } ?: break
                                if (sr.entryType == ScheduleRequirementType.RELATIVE_PAUSE && sr.matId != null && sr.durationMinutes != null) {
                                    q.add(sr to emptyList())
                                } else {
                                    if (requirementsCapacity[requiremetsGraph.getIndex(sr.id)] > 0) {
                                        val fights = st.flushCompletableFights(requiremetsGraph.getFightIdsForRequirement(sr.id))
                                        q.add(sr to fights)
                                    }
                                }
                            }
                            while (!q.isEmpty()) {
                                val req = q.poll()
                                if (req.first.entryType == ScheduleRequirementType.RELATIVE_PAUSE && !req.first.matId.isNullOrBlank() && req.first.durationMinutes != null
                                        && req.first.periodId == period.id) {
                                    if (matsToIds.containsKey(req.first.matId) && matsToIds.getValue(req.first.matId).periodId == period.id) {
                                        val mat = matsToIds.getValue(req.first.matId)
                                        log.info("Processing pause at mat ${req.first.matId}, period ${req.first.periodId} at ${mat.currentTime} for ${req.first.durationMinutes} minutes")
                                        val e = createRelativePauseEntry(req.first, mat.currentTime,
                                                mat.currentTime.plus(req.first.durationMinutes.toLong(), ChronoUnit.MINUTES))
                                        accumulator.scheduleEntries.add(e)
                                        mat.currentTime = mat.currentTime.plus(req.first.durationMinutes.toLong(), ChronoUnit.MINUTES)
                                    } else {
                                        log.warn("Relative pause ${req.first.id} not dispatched because either mat ${req.first.matId} not found or mat is from another period: ${matsToIds[req.first.matId]?.periodId}/${period.id}")
                                    }
                                    continue
                                }
                                log.info("Processing requirement: ${req.first.id}")
                                val capacity = requirementsCapacity[requiremetsGraph.getIndex(req.first.id)]
                                if (!req.first.matId.isNullOrBlank()) {
                                    if (matsToIds.containsKey(req.first.matId)) {
                                        val mat = matsToIds[req.first.matId]
                                                ?: error("No mat with id: ${req.first.matId}")
                                        if (mat.periodId == period.id) {
                                            req.second.forEach { fightId ->
                                                updateMatAndSchedule(requirementsCapacity, requiremetsGraph, req, accumulator, mat, fightId, st, period)
                                                fightsDispatched++
                                                log.debug("Dispatched $fightsDispatched fights")
                                            }
                                        } else {
                                            log.warn("Mat with id ${req.first.matId} is from a different period. Dispatching to the available mats for this period.")
                                            fightsDispatched = loadBalanceToMats(req, periodMats, requirementsCapacity, requiremetsGraph, accumulator, st, period, fightsDispatched)
                                        }
                                    } else {
                                        log.warn("No mat with id ${req.first.matId}")
                                    }
                                } else {
                                    fightsDispatched = loadBalanceToMats(req, periodMats, requirementsCapacity, requiremetsGraph, accumulator, st, period, fightsDispatched)
                                }

                                if (capacity > 0 && capacity == requirementsCapacity[requiremetsGraph.getIndex(req.first.id)]) {
                                    log.info("Could not dispatch any of $capacity fights from requirement ${req.first.id}. Moving it to unfinished.")
                                    unfinishedRequirements.offer(req.first)
                                } else if (requirementsCapacity[requiremetsGraph.getIndex(req.first.id)] > 0) {
                                    q.offer(req.first to st.flushCompletableFights(requiremetsGraph.getFightIdsForRequirement(req.first.id)))
                                }
                            }
                            if (onlyUnfinished && n == st.getNonCompleteCount()) {
                                if (n > 0) {
                                    log.error("No progress on unfinished requirements and no new ones... breaking out of the loop.")
                                }
                                break
                            }
                        }
                    }
                    if (st.getNonCompleteCount() > 0) {
                        log.warn("${st.getNonCompleteCount()} fights were not dispatched.")
                    }
                    Tuple3(accumulator.scheduleEntries, accumulator.matSchedules.filterNotNull(), accumulator.invalidFights.toSet())
                }
    }

    private fun loadBalanceToMats(req: Pair<ScheduleRequirementDTO, List<String>>, periodMats: List<InternalMatScheduleContainer>, requirementsCapacity: IntArray, requiremetsGraph: RequirementsGraph, accumulator: ScheduleAccumulator, st: StageGraph, period: PeriodDTO, fightsDispatched: Int): Int {
        var fightsDispatched1 = fightsDispatched
        req.second.forEach { fightId ->
            val mat = periodMats.minByOrNull { it.currentTime.toEpochMilli() }!!
            updateMatAndSchedule(requirementsCapacity, requiremetsGraph, req, accumulator, mat, fightId, st, period)
            fightsDispatched1++
            log.debug("Dispatched $fightsDispatched1 fights")
        }
        return fightsDispatched1
    }

    private fun updateMatAndSchedule(requirementsCapacity: IntArray,
                                     requiremetsGraph: RequirementsGraph,
                                     req: Pair<ScheduleRequirementDTO, List<String>>,
                                     accumulator: ScheduleAccumulator,
                                     mat: InternalMatScheduleContainer,
                                     fightId: String,
                                     st: StageGraph,
                                     period: PeriodDTO) {
        val duration = getFightDuration(st.getDuration(fightId), period.id)
        if (!pauses[mat.id].isNullOrEmpty() && mat.currentTime.toEpochMilli() + eightyPercentOfDurationInMillis(duration) >= pauses.getValue(mat.id)[0].startTime.toEpochMilli()) {
            val p = pauses.getValue(mat.id).removeAt(0)
            log.info("Processing a fixed pause, required: ${p.startTime}, actual: ${mat.currentTime}, mat: ${mat.id}, duration: ${p.durationMinutes}. Period starts: ${period.startTime}")
            val e = createFixedPauseEntry(p, p.startTime.plus(p.durationMinutes.toLong(), ChronoUnit.MINUTES))
            accumulator.scheduleEntries.add(e)
            mat.currentTime = mat.currentTime.plus(p.durationMinutes.toLong(), ChronoUnit.MINUTES)
        }
        requirementsCapacity[requiremetsGraph.getIndex(req.first.id)]--
        val e = accumulator.scheduleEntryFromRequirement(req.first, mat.currentTime, period.id)
        accumulator.scheduleEntries[e].fightIds += MatIdAndSomeId(mat.id, mat.currentTime, fightId)
        accumulator.scheduleEntries[e].categoryIds += st.getCategoryId(fightId)
        mat.fights.add(InternalFightStartTime(fightId, st.getCategoryId(fightId), mat.id, mat.totalFights++, mat.currentTime, accumulator.scheduleEntries[e].id, period.id))
        log.info("Period: ${period.id}, category: ${st.getCategoryId(fightId)}, fight: $fightId, starts: ${mat.currentTime}, mat: ${mat.id}, numberOnMat: ${mat.totalFights - 1}")
        mat.currentTime = mat.currentTime.plus(duration.toLong(), ChronoUnit.MINUTES)
        st.completeFight(fightId)
    }

    private val initialFightsByMats = mats.mapIndexed { i, mat ->
        val initDate = ZonedDateTime.ofInstant(startTime[mat.periodId]
                ?: error("No Start time for period ${mat.periodId}"), ZoneId.of(timeZone))
        InternalMatScheduleContainer(
                timeZone = timeZone,
                name = mat.name,
                id = mat.id ?: IDGenerator.createMatId(mat.periodId),
                fights = mutableListOf(),
                currentTime = initDate.toInstant(),
                totalFights = 0,
                matOrder = mat.matOrder ?: i,
                periodId = mat.periodId)
    }

    private fun supplier(): Supplier<ScheduleAccumulator> {
        return Supplier {
            log.info("Supplier.")
            ScheduleAccumulator(initialFightsByMats, competitionId)
        }
    }
}

File: saga-processor/src/main/scala/compman/compsrv/logic/service/schedule/ScheduleProducer.scala
package compman.compsrv.logic.service.schedule

import compman.compsrv.logic.service.fights.CanFail
import compman.compsrv.model.dto.dashboard.MatDescriptionDTO
import compman.compsrv.model.dto.schedule._
import compman.compsrv.model.extension._

import java.time.{Instant, ZonedDateTime, ZoneId}
import java.time.temporal.ChronoUnit
import java.util.UUID
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.util.Try
import cats.implicits._
import compman.compsrv.model.Errors

object ScheduleProducer {
  private def eightyPercentOfDurationInMillis(duration: Long): Long = duration * 8 / 10

  private def createPauseEntry(
    pauseReq: ScheduleRequirementDTO,
    startTime: Instant,
    endTime: Instant,
    pauseType: ScheduleEntryType
  ): ScheduleEntryDTO = {
    val entry = new ScheduleEntryDTO()
    entry.setId(pauseReq.getId)
    entry.setCategoryIds(Array.empty)
    entry.setFightIds(Array(
      new MatIdAndSomeId().setSomeId(pauseReq.getId).setMatId(pauseReq.getMatId).setStartTime(startTime)
    ))
    entry.setPeriodId(pauseReq.getPeriodId)
    entry.setStartTime(startTime)
    entry.setNumberOfFights(0)
    entry.setEntryType(pauseType)
    entry.setEndTime(endTime)
    entry.setDuration(pauseReq.getDurationSeconds)
    entry.setRequirementIds(Array(pauseReq.getId))
  }

  private def createFixedPauseEntry(fixedPause: ScheduleRequirementDTO, endTime: Instant) =
    createPauseEntry(fixedPause, fixedPause.getStartTime, endTime, ScheduleEntryType.FIXED_PAUSE)

  private def createRelativePauseEntry(requirement: ScheduleRequirementDTO, startTime: Instant, endTime: Instant) =
    createPauseEntry(requirement, startTime, endTime, ScheduleEntryType.RELATIVE_PAUSE)

  private def getFightDuration(duration: Int, riskCoeff: Int, timeBetweenFights: Int): Int = duration *
    (1 + 100 / riskCoeff) + timeBetweenFights

  def simulate(
    stageGraph: StageGraph,
    requiremetsGraph: RequirementsGraph,
    periods: List[PeriodDTO],
    req: List[ScheduleRequirementDTO],
    periodStartTime: Map[String, Instant],
    mats: List[MatDescriptionDTO],
    timeZone: String
  ): CanFail[(List[ScheduleEntryDTO], List[InternalMatScheduleContainer], Set[String])] = Try {
    val initialFightsByMats: List[InternalMatScheduleContainer] =
      createMatScheduleContainers(mats, periodStartTime, timeZone)
    val accumulator: ScheduleAccumulator = ScheduleAccumulator(initialFightsByMats)
    val pauses = req.filter { _.getEntryType == ScheduleRequirementType.FIXED_PAUSE }.groupBy { _.getMatId }.view
      .mapValues { e => e.sortBy { _.getStartTime.toEpochMilli() } }.mapValues(ArrayBuffer.from(_))
    val unfinishedRequirements = mutable.Queue.empty[ScheduleRequirementDTO]
    val matsToIds              = accumulator.matSchedules.groupMapReduce(_.id)(identity)((a, _) => a)
    val sortedPeriods          = periods.sortBy { _.getStartTime }
    val requirementsCapacity   = requiremetsGraph.requirementFightsSize

    def loadBalanceToMats(
      req: (ScheduleRequirementDTO, List[String]),
      periodMats: Seq[InternalMatScheduleContainer],
      requirementsCapacity: Array[Int],
      requiremetsGraph: RequirementsGraph,
      accumulator: ScheduleAccumulator,
      st: StageGraph,
      period: PeriodDTO
    ): StageGraph = {
      req._2.foldLeft(st) { (stageGraph, fightId) =>
        val mat = periodMats.minBy { _.currentTime.toEpochMilli() }
        updateMatAndSchedule(requirementsCapacity, requiremetsGraph, req, accumulator, mat, fightId, stageGraph, period)
      }
    }

    def updateMatAndSchedule(
      requirementsCapacity: Array[Int],
      requiremetsGraph: RequirementsGraph,
      req: (ScheduleRequirementDTO, List[String]),
      accumulator: ScheduleAccumulator,
      mat: InternalMatScheduleContainer,
      fightId: String,
      st: StageGraph,
      period: PeriodDTO
    ): StageGraph = {
      val duration = getFightDuration(
        st.getDuration(fightId).toInt,
        (period.getRiskPercent.doubleValue() * 100).toInt,
        period.getTimeBetweenFights
      )
      if (
        !(pauses(mat.id) == null || pauses(mat.id).isEmpty) &&
        mat.currentTime.toEpochMilli + eightyPercentOfDurationInMillis(duration) >= pauses(mat.id)(0)
          .getStartTime.toEpochMilli
      ) {
        val p = pauses(mat.id).remove(0)
        val e = createFixedPauseEntry(p, p.getStartTime.plus(p.getDurationSeconds.toLong, ChronoUnit.SECONDS))
        accumulator.scheduleEntries.append(e)
        mat.currentTime = mat.currentTime.plus(p.getDurationSeconds.toLong, ChronoUnit.SECONDS)
      }
      requirementsCapacity(requiremetsGraph.getIndex(req._1.getId).getOrElse(-1)) -= 1
      val e = accumulator.scheduleEntryFromRequirement(req._1, mat.currentTime, Option(period.getId))
      accumulator.scheduleEntries(e).addCategoryId(st.getCategoryId(fightId))
      mat.fights.append(InternalFightStartTime(
        fightId,
        st.getCategoryId(fightId),
        mat.id,
        mat.totalFights,
        mat.currentTime,
        accumulator.scheduleEntries(e).getId,
        period.getId
      ))
      mat.totalFights += 1
      mat.currentTime = mat.currentTime.plus(duration.toLong, ChronoUnit.SECONDS)
      StageGraph.completeFight(fightId, st)
    }

    def addRequirementsToQueue(
      queueToUpdate: mutable.Queue[(ScheduleRequirementDTO, List[String])],
      requirementsQueue: mutable.Queue[ScheduleRequirementDTO]
    ): Unit = {
      var i = 0
      while ((requirementsQueue.nonEmpty && i < initialFightsByMats.size) || unfinishedRequirements.nonEmpty) {
        val sr =
          if (unfinishedRequirements.nonEmpty) unfinishedRequirements.dequeue()
          else if (requirementsQueue.nonEmpty) {
            i += 1
            requirementsQueue.dequeue()
          } else { return }
        if (
          sr.getEntryType == ScheduleRequirementType.RELATIVE_PAUSE && sr.getMatId != null && sr.getDurationSeconds > 0
        ) { queueToUpdate.append((sr, List.empty)) }
        else {
          val i1 = requiremetsGraph.getIndex(sr.getId).getOrElse(-1)
          if (i1 >= 0 && requirementsCapacity(i1) > 0) {
            val fights = stageGraph.flushCompletableFights(requiremetsGraph.getFightIdsForRequirement(sr.getId))
            queueToUpdate.append((sr, fights))
          }
        }
      }

    }

    def dispatchFightsFromQueue(
      period: PeriodDTO,
      periodMats: mutable.Seq[InternalMatScheduleContainer],
      q: mutable.Queue[(ScheduleRequirementDTO, List[String])]
    ): Unit = {
      while (q.nonEmpty) {
        val req = q.dequeue()
        if (
          req._1.getEntryType == ScheduleRequirementType.RELATIVE_PAUSE && req._1.getMatId != null &&
          req._1.getDurationSeconds != 0 && req._1.getPeriodId == period.getId
        ) {
          if (matsToIds.contains(req._1.getMatId) && matsToIds(req._1.getMatId).periodId == period.getId) {
            val mat = matsToIds(req._1.getMatId)
            val e = createRelativePauseEntry(
              req._1,
              mat.currentTime,
              mat.currentTime.plus(req._1.getDurationSeconds.toLong, ChronoUnit.SECONDS)
            )
            accumulator.scheduleEntries.append(e)
            mat.currentTime = mat.currentTime.plus(req._1.getDurationSeconds.toLong, ChronoUnit.SECONDS)
          }
        } else {
          val ind      = requiremetsGraph.getIndex(req._1.getId)
          val capacity = requirementsCapacity(ind.getOrElse(-1))
          if (req._1.getMatId != null) {
            if (matsToIds.contains(req._1.getMatId)) {
              val mat = matsToIds(req._1.getMatId)
              if (mat.periodId == period.getId) {
                loadBalanceToMats(
                  req,
                  Seq(mat),
                  requirementsCapacity,
                  requiremetsGraph,
                  accumulator,
                  stageGraph,
                  period
                )
              } else {
                loadBalanceToMats(
                  req,
                  periodMats.toSeq,
                  requirementsCapacity,
                  requiremetsGraph,
                  accumulator,
                  stageGraph,
                  period
                )
              }
            }
          } else {
            loadBalanceToMats(
              req,
              periodMats.toSeq,
              requirementsCapacity,
              requiremetsGraph,
              accumulator,
              stageGraph,
              period
            )
          }

          if (capacity > 0 && capacity == requirementsCapacity(requiremetsGraph.getIndexOrMinus1(req._1.getId))) {
            unfinishedRequirements.enqueue(req._1)
          } else if (requirementsCapacity(requiremetsGraph.getIndexOrMinus1(req._1.getId)) > 0) {
            q.enqueue(
              (req._1, stageGraph.flushCompletableFights(requiremetsGraph.getFightIdsForRequirement(req._1.getId)))
            )
          }
        }
      }
    }

    sortedPeriods.foreach { period =>
      val periodMats = accumulator.matSchedules.filter { _.periodId == period.getId }
      unfinishedRequirements.foreach { r =>
        stageGraph.flushNonCompletedFights(requiremetsGraph.getFightIdsForRequirement(r.getId)).foreach { it =>
          accumulator.invalidFights.add(it)
        }
      }
      val q: mutable.Queue[(ScheduleRequirementDTO, List[String])] = mutable.Queue.empty
      val rq: mutable.Queue[ScheduleRequirementDTO] = mutable.Queue
        .from(requiremetsGraph.orderedRequirements.filter { it => it.getPeriodId == period.getId })
      var dispatchSuccessful = true
      while ((rq.nonEmpty || unfinishedRequirements.nonEmpty) && dispatchSuccessful) {
        val n              = stageGraph.getNonCompleteCount
        val onlyUnfinished = rq.isEmpty
        addRequirementsToQueue(q, rq)
        dispatchFightsFromQueue(period, periodMats, q)
        dispatchSuccessful = onlyUnfinished && n == stageGraph.getNonCompleteCount
      }
    }
    (
      accumulator.scheduleEntries.toList,
      accumulator.matSchedules.filter(_ != null).toList,
      accumulator.invalidFights.toSet
    )}.toEither.leftMap(t => Errors.InternalError(t.getMessage))

  private def createMatScheduleContainers(
    mats: List[MatDescriptionDTO],
    periodStartTime: Map[String, Instant],
    timeZone: String
  ) = {
    val initialFightsByMats = mats.zipWithIndex.map { case (mat, i) =>
      val initDate = ZonedDateTime.ofInstant(periodStartTime(mat.getPeriodId), ZoneId.of(timeZone))
      InternalMatScheduleContainer(
        timeZone = timeZone,
        name = mat.getName,
        id = Option(mat.getId).getOrElse(UUID.randomUUID().toString),
        fights = ArrayBuffer.empty,
        currentTime = initDate.toInstant,
        totalFights = 0,
        matOrder = Option(mat.getMatOrder).map(_.toInt).getOrElse(i),
        periodId = mat.getPeriodId
      )
    }
    initialFightsByMats
  }
}
